rasa/engine/graph.py

```diff
@@ -0,0 +1,196 @@
+    def supported_languages(self) -> List[Text]:
+        """Determines which languages this component can work with."""
+        ...
+
+    @abstractmethod
+    def required_packages(self) -> List[Text]:
+        """Any extra python dependencies required for this component to run."""
+        ...
+
+
+@dataclass
+class ExecutionContext:
+    """Holds information about a single graph run."""
+
+    graph_schema: GraphSchema = field(repr=False)
+    model_id: Text
+    diagnostic_data: bool = False
+    is_finetuning: bool = False
+
+
+class GraphNode:
+    """Instantiates and runs a `GraphComponent` within a graph.
+
+    A `GraphNode` is a wrapper for a `GraphComponent` that allows it to be executed
+    In the context of a graph. It is responsible for instantiating the component at the
+    correct time, collecting the inputs from the parent nodes, running the run function
+    of the component and passing the output onwards.
+    """
+
+    def __init__(
+        self,
+        node_name: Text,
+        component_class: Type[GraphComponent],
+        constructor_name: Text,
+        component_config: Dict[Text, Any],
+        fn_name: Text,
+        inputs: Dict[Text, Text],
+        eager: bool,
+        execution_context: ExecutionContext,
+    ) -> None:
+        """Initializes `GraphNode`.
+
+        Args:
+            node_name: The name of the node in the schema.
+            component_class: The class to be instantiated and run.
+            constructor_name: The method used to instantiate the component.
+            component_config: Config to be passed to the component.
+            fn_name: The function to be run when the node executes.
+            inputs: A map from input name to parent node name that provides it.
+            eager: Determines if the node is instantiated right away, or just before
+                being run.
+            execution_context: Information about the current graph run.
+        """
+        self._node_name: Text = node_name
+        self._component_class: Type[GraphComponent] = component_class
+        self._constructor_name: Text = constructor_name
+        self._constructor_fn: Callable = getattr(
+            self._component_class, self._constructor_name
+        )
+        self._component_config: Dict[Text, Any] = {
+            **self._component_class.default_config,
+            **component_config,
+        }
+        self._fn_name: Text = fn_name
+        self._fn: Callable = getattr(self._component_class, self._fn_name)
+        self._inputs: Dict[Text, Text] = inputs
+        self._eager: bool = eager
+        self._execution_context: ExecutionContext = execution_context
+
+        self._component: Optional[GraphComponent] = None
+        if self._eager:
+            self._load_component()
+
+    def _load_component(
+        self, additional_kwargs: Optional[Dict[Text, Any]] = None
+    ) -> None:
+        kwargs = additional_kwargs if additional_kwargs else {}
+
+        logger.debug(
```

Comment:
I was tempted to create a custom logger in the graph node which automatically adds some formatting and information - do you think it's worth it?

Reply:
Great idea! I'd take a note of it in the implementation proposal and defer it to later. I'd keep the `engine` PR as small and simple as possible, merge to `main` and then add the bells and whistles bit by bit